dashboard-frontend/src/fixtures/FakeStudyBuilder.ts

```diff
@@ -0,0 +1,306 @@
+import Chance from 'chance';
+import {
+  calculateAverageDeviation,
+  calculatePercentage,
+  round,
+} from '../helpers/numbers';
+import { createSlugFor } from '../helpers/strings';
+import { lastValue } from '../helpers/arrays';
+import { Currency } from '../types/shared';
+import { StratumProgress, Study, StudyProgress } from '../types/study';
+
+const chance = Chance();
+
+export class FakeStudyBuilder {
+  creationDate?: number;
+  desiredParticipants?: number;
+  currency?: Currency;
+  numOfDifferentStrata?: number;
+  desiredParticipantsPerStrata?: number;
+  totalHoursOfData?: number;
+
+  public setCreationDate(date: number) {
+    this.creationDate = date;
+    return this;
+  }
+
+  public setDesiredParticipants(participants: number) {
+    this.desiredParticipants = participants;
+    return this;
+  }
+
+  public setCurrency(currency: Currency) {
+    this.currency = currency;
+    return this;
+  }
+
+  public setStrata(options: {
+    numOfDifferentStrata: number;
+    desiredParticipantsPerStrata: number;
+  }) {
+    this.numOfDifferentStrata = options.numOfDifferentStrata;
+    this.desiredParticipantsPerStrata = options.desiredParticipantsPerStrata;
+    return this;
+  }
+
+  public generateHourlyDataFor(options: { hours: number }) {
+    this.totalHoursOfData = options.hours;
+    return this;
+  }
+
+  public build() {
+    const numOfDifferentStrata = this.getNumOfDifferentStrata();
+    const totalHoursOfData = this.getTotalHoursOfData();
+
+    let study = createInitialStudyData({
+      creationDate: this.getCreationDate(),
+      desiredParticipants: this.getDesiredParticipants(),
+      numOfDifferentStrata,
+      currency: this.getCurrency(),
+      desiredParticipantsPerStrata: this.getDesiredParticipantsPerStrata(),
+    });
+
+    for (let hours = 1; hours <= totalHoursOfData; hours++) {
+      const lastStudyProgress = lastValue(study.studyProgressList);
+      const oneHourInMilliseconds = 3600000;
+
+      study = updateStudyData(study, {
+        datetime: lastStudyProgress.datetime + oneHourInMilliseconds,
+        maxOfNewParticipants: this.getDesiredParticipants() / totalHoursOfData,
+        numOfDifferentStrata: this.getNumOfDifferentStrata(),
+      });
+    }
+
+    return study;
+  }
+
+  private getCreationDate() {
+    if (!this.creationDate) {
+      throw this.requiredCallError('setCreationDate');
+    }
+
+    return this.creationDate;
+  }
+
+  private getDesiredParticipants() {
```

Comment:
I have the sense you've been indoctrinated into Java somehow.

We'll need to unindoctrinate you! 

All you're doing here is validating a config. Rather than create methods for all the config params (Java's "Builder" pattern), you can just create a "config" object (this is super well supported in JS, this is why ES6 went to all the crazy trouble to support defaults inside objects) or just use named params with types or whatever. If you have a config, you just run a single loop over the config and check for required fields (or just json schema or, actually, use your type system to catch missing fields at compile time, given that you have a type system, no?). 

That will remove about 120 lines to this file and get you the same result ;)

Reply:
Yes, we can get rid of it 👍

I ended up reaching for this pattern to make it more expressive and clear on the callsite, but you're right it ended up being an overkill. 😞

I'll refactor it into a function that accepts a config object and I'll rely on the type system to ensure that all the required config is provided.